### PR TITLE
Create op builders (1.)

### PR DESCRIPTION
**Pull request by gchinora**
First PR related to op builders. 

The whole purpose of the op-builders is to factor out the main functionalities behind the current onnx parsers and put them in a separate set of files (under the folder ./src/op/builder/ ) so they could be reused in different parts of the application. That is this is mainly a refactoring task; the passing of the already existing unit and verify tests should be the proof that the refactoring left the integrity of mgr intact.

- related issue: https://github.com/migraphx-benchmark/AMDMIGraphX/issues/203
- covered onnx parsers in this PR:  einsum, gemm, mean_variance_normalization, gelu, batchnorm, celu

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -79,7 +79,7 @@ docs/html
 # JetBrains config directories (ignoring symlinks)
 .idea/
 cmake-build*/
-build*/
```
Comment (CharlieL7) [inline .gitignore:82]: Is this a debug leftover?

Comment (apwojcik) [reply .gitignore:82]: @CharlieL7 No, please keep it as is.

**Comment by kahmed10** [comment]
As a general note, please update license files to 2025.

File: src/op/builder/gemm.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    insert(module& m, instruction_ref ins, const std::vector<instruction_ref>& args) const
+    {
+       auto x      = args[0];
+       auto x_type = x->get_shape().type();
+       auto alpha_lit = m.add_literal(migraphx::literal{migraphx::shape{x_type}, {alpha}});
+       auto mul_alpha = insert_common_op(m, ins, make_op("mul"), {alpha_lit, x});
+       auto sigmoid = m.insert_instruction(ins, migraphx::make_op("sigmoid"), mul_alpha);
+       return {insert_common_op(m, ins, make_op("mul"), {x, sigmoid})};
+    }
+};
```
Comment (CharlieL7) [inline src/op/builder/gemm.cpp:59]: why is this in the gemm builder?

File: src/op/builder/op_builder.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+std::vector<instruction_ref> insert(const std::string& name,
+                                    module& m,
+                                    instruction_ref ins,
+                                    const std::vector<instruction_ref>& args,
+                                    const value& options)
+{
+    return at(builder_map(), name, "Builder not found: " + name)(m, ins, args, options);
+}
```
Comment (CharlieL7) [inline src/op/builder/op_builder.cpp:53]: How will this handle operators with module inputs? Like the `If` operator: https://onnx.ai/onnx/operators/onnx__If.html. Current parse code is here: src/onnx/parse_if.cpp. It looks like we would have to have another option to also pass the module inputs.

**Review by CharlieL7 — CHANGES_REQUESTED** [review]

File: src/onnx/parse_batchnorm.cpp [hunk 4]
```diff
@@ -37,61 +36,15 @@ struct parse_batchnorm : op_parser<parse_batchnorm>
     instruction_ref parse(const op_desc& /*opd*/,
                           const onnx_parser& parser,
                           const onnx_parser::node_info& info,
-                          std::vector<instruction_ref> args) const
+                          const std::vector<instruction_ref>& args) const
     {
         float epsilon = 1e-5f;
```
Comment (CharlieL7) [inline src/onnx/parse_batchnorm.cpp:41]: The default attribute value is set in parse_... and in the op builder. This might lead to some unintuitive behavior should someone try to change it in the op builder. Probably better to have a conditional here and use the default in the op builder.

File: test/onnx/include/onnx_test_utils.hpp [hunk 5]
```diff
@@ -329,6 +328,8 @@ inline void mvn_n_rank_test(std::vector<int64_t> axes,
     auto divisor  = add_common_op(*mm, make_op("add"), {std, epsilon});
     add_common_op(*mm, make_op("div"), {dividend, divisor});
 
+    p.get_main_module()->sort();
+    prog.get_main_module()->sort();
```
Comment (CharlieL7) [inline test/onnx/include/onnx_test_utils.hpp:332]: Adding sorting when before we did not need it. Why has the ordering changed?

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+template <class T>
+void register_builder()
+{
+    builder_func f = [](module& m,
+                        instruction_ref ins,
+                        const std::vector<instruction_ref>& args,
+                        std::vector<module_ref> module_args,
+                        const value& options) {
+        auto x = from_value<T>(options);
+        return x.insert(insert_params{m, ins, args, module_args});
```
Comment (pfultz2) [inline src/op/builder/include/migraphx/op/builder/op_builder.hpp:65]: This should call

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+MIGRAPHX_EXPORT void register_builder(const std::string& name, builder_func f);
+
+struct insert_params
+{
+    module& m;
+    instruction_ref ins;
+    std::vector<instruction_ref> args;
+    std::vector<module_ref> module_args;
+};
```
Comment (pfultz2) [inline src/op/builder/include/migraphx/op/builder/op_builder.hpp]: Remove this struct, these can just be passed as parameters.

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp (hunk 6 above)
Comment (pfultz2) [inline src/op/builder/include/migraphx/op/builder/op_builder.hpp]: Pass the parameters directly, but you need to check if it accepts `module_args`.

**Review by pfultz2 — CHANGES_REQUESTED** [review]

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+namespace builder {
+
+using builder_func =
+    std::function<std::vector<instruction_ref>(module& m,
+                                               instruction_ref ins,
+                                               const std::vector<instruction_ref>& args,
+                                               std::vector<module_ref> module_args,
```
Comment (pfultz2) [inline src/op/builder/include/migraphx/op/builder/op_builder.hpp]: This should be const ref.

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp (hunk 6 above)
Comment (pfultz2) [reply src/op/builder/include/migraphx/op/builder/op_builder.hpp]: You can use an `invoke_builder` function like this to handle this case:

```cpp
template<class T>
auto invoke_builder(module& m,
                    instruction_ref ins,
                    const std::vector<instruction_ref>& args,
                    std::vector<module_ref> module_args,
                    const value& options) -> decltype(T{}.insert(m, ins, args, module_args)) {
        auto x = from_value<T>(options);
        return x.insert(m, ins, args, module_args);
}


template<class T>
auto invoke_builder(module& m,
                    instruction_ref ins,
                    const std::vector<instruction_ref>& args,
                    std::vector<module_ref> module_args,
                    const value& options) -> decltype(T{}.insert(m, ins, args)) {
        if(module_args.empty())
            MIGRAPHX_THROW("Module args should be empty");
        auto x = from_value<T>(options);
        return x.insert(m, ins, args);
}
```

File: src/op/builder/mean_variance_normalization.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        return pack(f(self.axes, "axes"));
+    }
+
+    static std::string name() { return "mean_variance_normalization"; }
+
+    std::vector<instruction_ref> insert(insert_params params)
+    {
+        return insert(params.m, params.ins, params.args);
+    }
```
Comment (pfultz2) [inline src/op/builder/mean_variance_normalization.cpp]: Remove this method.

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+{
+    template <class T>
+    static void apply()
+    {
+        register_builder<T>();
+    }
+};
+
+template <class T>
+using op_builder = auto_register<register_builder_action, T>;
```
Comment (pfultz2) [inline src/op/builder/include/migraphx/op/builder/op_builder.hpp]: We should make `op_builder` an actual class with a default name method:

```cpp
template <class T>
struct op_builder : register_op_builder<T>
{
    static std::string name()
    {
        static const std::string& name = get_type_name<T>();
        return name.substr(name.rfind("::") + 2);
    }
};
```

File: src/op/builder/mean_variance_normalization.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+{
+    std::vector<int64_t> axes{0, 2, 3};
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.axes, "axes"));
+    }
+
+    static std::string name() { return "mean_variance_normalization"; }
```
Comment (pfultz2) [inline src/op/builder/mean_variance_normalization.cpp]: Remove this, since it is the same as the base class.

File: test/include/pointwise.hpp [hunk 12]
```diff
@@ -65,7 +65,7 @@ migraphx::instruction_ref add_pointwise(migraphx::program& p,
                                         const std::vector<migraphx::instruction_ref>& inputs,
                                         const F& f)
 {
-    return add_pointwise(p, p.get_main_module(), name, std::move(inputs), std::move(f));
+    return add_pointwise(p, p.get_main_module(), name, inputs, std::move(f));
```
[the remark this answers is not in the document]
Comment (gchinora) [reply test/include/pointwise.hpp:68]: Running "make -j64 -k onnx-proto tf-proto tidy" complained about this. So I'll keep the std::move removed

File: src/op/builder/gelu.cpp [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        return insert(params.m, params.ins, params.args);
+    }
+
+    std::vector<instruction_ref>
+    insert(module& m, instruction_ref ins, const std::vector<instruction_ref>& args) const
+    {
+       auto x      = args[0];
+       auto x_type = x->get_shape().type();
+       auto alpha_lit = m.add_literal(migraphx::literal{migraphx::shape{x_type}, {alpha}});
+       auto mul_alpha = insert_common_op(m, ins, make_op("mul"), {alpha_lit, x});
```
[the remark this answers is not in the document]
Comment (gchinora) [reply src/op/builder/gelu.cpp]: The suggested modification causes compile-time error. Won't apply!

File: src/op/builder/gelu.cpp [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+       auto sigmoid = m.insert_instruction(ins, migraphx::make_op("sigmoid"), mul_alpha);
+       return {insert_common_op(m, ins, make_op("mul"), {x, sigmoid})};
```
[the remark this answers is not in the document]
Comment (gchinora) [reply src/op/builder/gelu.cpp]: The suggested modification causes compile-time error. Won't apply!

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp (hunk 6 above)
Comment (gchinora) [reply src/op/builder/include/migraphx/op/builder/op_builder.hpp]: in the if statement you wanted to put a "not" there, didn't you?

File: src/op/builder/gelu.cpp [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+
+    std::vector<instruction_ref>
+    insert(module& m, instruction_ref ins, const std::vector<instruction_ref>& args) const
+    {
+        auto x = args[0];
+        size_t last_dim_size = x->get_shape().lens().back();
+        if (last_dim_size < 2 or last_dim_size % 2 != 0)
+            MIGRAPHX_THROW("PARSE_GELU: BiasSplitGelu must have even last dimension which is >= 2");
```
Comment (pfultz2) [inline src/op/builder/gelu.cpp]: The error message should refer to the builder and not the onnx operator: `MIGRAPHX_THROW("gelu_split: must have even last dimension which is >= 2")`

File: src/op/builder/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#####################################################################################
+
+file(GLOB BUILDER_SRCS CONFIGURE_DEPENDS *.cpp)
+add_library(migraphx_builders ${BUILDER_SRCS})
```
Comment (pfultz2) [inline src/op/builder/CMakeLists.txt]: Dont make this a seperate library, just add the sources to the migraphx library: `target_sources(migraphx ${BUILDER_SRCS})`. Of course, since it is just two lines, then this should probably go into the src/CMakeLists.txt.

File: src/op/builder/include/migraphx/op/builder/op_builder.hpp (hunk 6 above)
Comment (pfultz2) [reply src/op/builder/include/migraphx/op/builder/op_builder.hpp]: > in the if statement you wanted to put a "not" there, didn't you?

Ah yes.

**Comment by causten** [comment]
@gchinora Please add some additional text in the PR that describes why.  This will allow reviewers to have better context.

**Review by CharlieL7 — APPROVED** [review]
LGTM

**Comment by lakhinderwalia** [comment]
1. I could identify here the need for a code reorg, for `gelu`.
2. For the other op(s), is this just a basically the same code being moved to different sub-dir? Where is the re-use, please?
3. Why is the total code (number of lines) increasing for reuse? This PR shows a net increase of 600 lines -- that is way more than the ~ 25 * 9 lines of the boilerplate copyright header(s).

Thanks.

**Comment by CharlieL7** [comment]
> 1. I could identify here the need for a code reorg, for `gelu`.
> 
>     2. For the other op(s), is this just a basically the same code being moved to different sub-dir? Where is the re-use, please?
> 
>     3. Why is the total code (number of lines) increasing for reuse? This PR shows a net increase of 600 lines -- that is way more than the ~ 25 * 9 lines of the boilerplate copyright header(s).
> 
> 
> Thanks.

2: The plan is to use these builders for parsing Tensorflow models and TorchMIGraphX. We basically re-implement many of these parsers in both those situations. Would be a separate PR to update those parsing.

3: Added code because the onnx parsers for each op still exists, just calls into the op builder now. Additionally there's some boilerplate for the op builder itself.
